lib/cocoapods/downloader/cache.rb

```diff
@@ -69,6 +69,60 @@ def cache_descriptors_per_pod
         end
       end
 
+      # Convenience method for acquiring a shared lock to safely read from the
+      # cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [Void]
+      #
+      def self.read_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_SH, block)
+      end
+
+      # Convenience method for acquiring an exclusive lock to safely write to
+      # the cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [Void]
+      #
+      def self.write_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_EX, block)
+      end
+
+      # Creates a .lock file at `location`, aquires a lock of type
+      # `lock_type`, and executes `block` while holding on to that lock. The
+      # block is passed the location for convenience.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [locking_constant] lock_type
+      #         the type of lock, either exclusive (File::LOCK_EX) or shared
+      #         (File::LOCK_SH).
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [Void]
+      #
+      def self.lock(location, lock_type, block = Proc.new)
+        lockfile = "#{location}.lock"
```

Comment:
is this file removed afterwords?

Reply:
It isn't, and it wasn't in #6535 either. I was experimenting with that, but I'm not sure how to ensure that it's safe to delete the file and that no other process is waiting to lock on that file. I didn't get it to work reliably, and I thought it probably doesn't matter much anyway. It's empty files that don't take up almost any space in a directory that isn't supposed to be read by humans, so it's hard to imagine that it would really bother anyone.

If you know how to safely delete a lock file without messing with other process' locking attempts, I would be happy to add that. :)

Reply:
Can we possibly use the `tmp` folder that the OS can then handle by itself?

Reply:
Maybe use `Tempfile.new`?

Reply:
`Tempfile.new` wouldn't work since it creates a new file with a random name each time, so each process would get its own file and would hence not use the same lock.

Reply:
Since you seem quite keen, I did some more research. There is a potential solution here:
https://www.ruby-forum.com/t/how-do-you-use-flock-and-clean-up-lock-files/64817/7

It seems a bit convoluted, but the logic looks sound. After acquiring the lock, you check if the file still exists on the filesystem and if it's the same file that you currently have a lock on, otherwise you abort and try again. Once you have a lock on a file that exists on the filesystem, you do your business and then you delete it before releasing the lock. Any processes that were waiting will detect that the file they got the lock for doesn't exist anymore and and try again as described above.

Anyway, it's getting quite late in my timezone, so I will probably not try it out until tomorrow.

Reply:
Actually, that logic doesn't consider shared locks, which I use for reading the cache. It gets quite complicated to accomodate those, since you need an exclusive lock to be sure that it's safe to delete the file. Unless you think it's absolutely necessary, I'd rather leave it as it is. Also, from a performance point of view, not deleting and recreating the lock file all the time should be advantageous.

Reply:
@dnkoutso I changed it so that the lock files are deleted after use, but now Inch doesn't like my documentation. No clue why. I added more documentation and somehow it's not good enough. Changing it back to what it was doesn't help, but deleting the documentation makes Inch happy. Seems like Inch is pretty useless. Any clue how I can keep the documentation and keep Inch happy?